core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java

```diff
@@ -342,54 +355,23 @@ private void trackState(IracManagerKeyState state) {
    private CompletionStage<Void> run() {
       // this run on a blocking thread!
       if (log.isTraceEnabled()) {
-         log.tracef("[IRAC] Sending keys to remote site(s). Has clear? %s, keys: %s", hasClear, updatedKeys.keySet());
+         log.tracef("[IRAC] Sending keys to remote site(s). Is clear? %s, keys: %s", hasClear, Util.toStr(updatedKeys.keySet()));
       }
       if (hasClear) {
          // clear doesn't work very well with concurrent updates
          // let's block all updates until the clear is applied everywhere
          return sendClearUpdate();
       }
 
-      for (IracManagerKeyState state : updatedKeys.values()) {
-         if (!state.canSend()) {
-            continue;
-         }
-         DistributionInfo dInfo = getDistributionInfo(state.getSegment());
-         if (!dInfo.isPrimary()) {
-            state.retry();
-            continue;
-         } else if (!dInfo.isWriteOwner()) {
-            // topology changed! cleanup the key
-            state.discard();
-            continue;
-         } else if (!dInfo.isReadOwner()) {
-            // state transfer in progress (we are a write owner but not a read owner)
-            // we need the data in DataContainer and CacheLoaders, so we must wait until we
-            // receive the key or will end up sending a remove update.
-            // when the new topology arrives, this will be triggered again
-            state.retry();
-            continue;
-         }
-
-         fetchEntry(state.getKey(), dInfo.segmentId()).thenApply(
-               lEntry -> lEntry == null ? buildRemoveCommand(state) : commandsFactory.buildIracPutKeyCommand(lEntry))
-               .thenAccept(cmd -> {
-                  if (cmd == null) { // only buildRemoveCommand can return null if tombstone is missing
-                     log.sendFailMissingTombstone(Util.toStr(state.getKey()));
-                     // avoid retrying
-                     onResponseCompleted(state, IracResponseCollector.Result.OK);
-                     return;
-                  }
-                  sendCommandToAllBackups(cmd, state, this);
-               })
-               .exceptionally(throwable -> {
-                  state.retry();
-                  onUnexpectedThrowable(throwable);
-                  return null;
-               });
-
-      }
-      return CompletableFutures.completedNull();
+      CompletableFutureObservable cfo = new CompletableFutureObservable();
+      Flowable.fromIterable(updatedKeys.values())
+            .filter(this::canStateBeSent)
+            .concatMapSingle(this::fetchEntry)
+            .filter(Objects::nonNull)
+            .buffer(batchSize)
+            .concatMapCompletable(this::sendUpdateBatch)
+            .subscribe(cfo);
+      return cfo;
```

Comment:
```suggestion
      return Flowable.fromIterable(updatedKeys.values())
            .filter(this::canStateBeSent)
            .concatMapMaybe(this::fetchEntry)
            .buffer(batchSize)
            .concatMapCompletable(this::sendUpdateBatch)
            .onErrorComplete(t -> {
               onUnexpectedThrowable(t);
               return true;
            })
            .toCompletionStage(null);
```
And remove the observable class at the end of the file.

Note this also requires the changes to change from Single to Maybe for fetchEntry.